File: go/bank-account/bank_account.go
package account

import "sync"

const testVersion = 1

type Account struct {
	balance int64
	closed  bool
	mutex   *sync.Mutex
}

type AccountIf interface {
	Close() (int64, bool)
	Balance() (int64, bool)
	Deposit(int64) (int64, bool)
}

func (a *Account) Close() (int64, bool) {
	a.mutex.Lock()
	if a.closed {
		a.mutex.Unlock()
		return 0, false
	}
	c := a.balance
	a.balance = 0
	a.closed = true
	a.mutex.Unlock()
	return c, true
}

func (a *Account) Balance() (int64, bool) {
	if a.closed {
		return 0, false
	}
	return a.balance, true
}

func (a *Account) Deposit(deposit int64) (int64, bool) {
	if a.closed {
		return 0, false
	}

	a.mutex.Lock()
	newBalance := a.balance + deposit
	if newBalance >= 0 {
		a.balance = newBalance
		a.mutex.Unlock()
		return a.balance, true
	}
	a.mutex.Unlock()
	return 0, false
}

func Open(deposit int64) *Account {
	if deposit < 0 {
		return nil
	}
	return &Account{deposit, false, &sync.Mutex{}}
}

File: python/bank-account/bank_account.py
from threading import Lock

class BankAccount(object):
    def __init__(self):
        self.balance = None
        self.lock = Lock()

    def get_balance(self):
        if self.balance is None:
            raise ValueError("Closed account")
        return self.balance

    def open(self):
        self.balance = 0

    def deposit(self, amount):
        try:
            self.lock.acquire()
            if self.balance is None:
                raise ValueError("Closed account")
            if amount <= 0:
                raise ValueError("Invalid deposit amount")
            self.balance += amount
        finally:
            self.lock.release()

    def withdraw(self, amount):
        try:
            self.lock.acquire()
            print(self.balance)
            if self.balance is None:
                raise ValueError("Closed account")
            if amount <= 0:
                raise ValueError("Invalid withdraw amount")
            if amount > self.balance:
                raise ValueError("Insufficent balance")
            self.balance -= amount
        finally:
            self.lock.release()

    def close(self):
        self.balance = None
